Add global hasher

masonforest opened the pull request:
> 😄 you might want to take a look at:
>Add a 'perl hash' implementation to spinoso-string #1450
>https://twitter.com/artichokeruby/status/1444089207547330562


Awesome! It's like we're Indiana Jonesing hashing algorithms haha. Let me know if this needs anything else!

Closes #1404

dependabot[bot] committed:
Bump rubocop from 1.22.0 to 1.22.3

Bumps [rubocop](https://github.com/rubocop/rubocop) from 1.22.0 to 1.22.3.
- [Release notes](https://github.com/rubocop/rubocop/releases)
- [Changelog](https://github.com/rubocop/rubocop/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rubocop/rubocop/compare/v1.22.0...v1.22.3)

---
updated-dependencies:
- dependency-name: rubocop
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

lopopolo committed:
Run cargo update in root workspace

```console
$ cargo update
    Updating crates.io index
    Updating cc v1.0.70 -> v1.0.71
    Updating clipboard-win v4.2.1 -> v4.2.2
    Updating libc v0.2.103 -> v0.2.106
    Updating onig v6.2.0 -> v6.3.1
    Updating onig_sys v69.7.0 -> v69.7.1
    Updating pkg-config v0.3.20 -> v0.3.22
    Updating ppv-lite86 v0.2.10 -> v0.2.15
    Updating proc-macro2 v1.0.29 -> v1.0.32
    Updating quote v1.0.9 -> v1.0.10
```

lopopolo committed:
Run cargo update in fuzz workspace

```console
$ cargo update
    Updating crates.io index
    Updating cc v1.0.70 -> v1.0.71
    Updating libc v0.2.103 -> v0.2.106
    Updating onig v6.2.0 -> v6.3.1
    Updating onig_sys v69.7.0 -> v69.7.1
    Updating pkg-config v0.3.20 -> v0.3.22
    Updating ppv-lite86 v0.2.10 -> v0.2.15
    Updating proc-macro2 v1.0.29 -> v1.0.32
    Updating quote v1.0.9 -> v1.0.10
```

lopopolo committed:
Run cargo update in spec-runner workspace

```console
$ cargo update
    Updating crates.io index
    Updating cc v1.0.70 -> v1.0.71
    Updating libc v0.2.103 -> v0.2.106
    Updating onig v6.2.0 -> v6.3.1
    Updating onig_sys v69.7.0 -> v69.7.1
    Updating pkg-config v0.3.20 -> v0.3.22
    Updating ppv-lite86 v0.2.10 -> v0.2.15
    Updating proc-macro2 v1.0.29 -> v1.0.32
    Updating quote v1.0.9 -> v1.0.10
    Updating syn v1.0.77 -> v1.0.81
```

lopopolo committed:
Bump onig version constraint to 6.3

lopopolo committed:
bundle exec rake lint:rubocop:auto_correct

masonforest committed:
Add a Global Hasher

masonforest committed:
Make artichoke_core::hash::BuildHasher generic

.. over std::hash::BuildHasher

Other changes:
* Return a referece from BuildHasher#build_hasher

lopopolo commented on artichoke-core/src/hash.rs:
```diff
@@ -0,0 +1,9 @@
+//! Build hashers and hash values
+
+/// A trait for creating instances of RandomState
+pub trait BuildHasher<B: std::hash::BuildHasher> {
+    /// Concrete error type for errors encountered when outputting hash errors.
+    type Error;
```
oh sorry I may have led you down a confusing path here.

Adding a generic parameter to the trait means that `BuildHasher` can be implemented multiple times for a given interpreter with multiple `BuildHasher` type params. This isn't quite the API I was hoping for.

Instead, like we have the associated `Error` type here, implementors of `BuildHasher` should declare exactly 1 `BuildHasher` type they use.

Something like this:

```suggestion
pub trait BuildHasher {
    /// Concrete error type for errors encountered when outputting hash errors.
    type Error;
    
    /// Concrete build hasher type.
    type BuildHasher: core::hash::BuildHasher;
```

This also uses the `BuildHasher` import via `core` so `artichoke-core` continues to be compilable with no-std.

This structure with associated types is used pretty heavily in `artichoke-core`, for example the PRNG in the `Prng` trait:

https://github.com/artichoke/artichoke/blob/1c1702894e6ef237b8d7644c16f6b1bc9ae5f5e7/artichoke-core/src/prng.rs#L10-L11

lopopolo replied:
another reason to prefer an associated item here:

If I'm building part of the standard library that wants to generically constrain the Artichoke interpreter to include a hasher, most of the time I don't care what `BuildHasher` is used, just that the trait is implemented.

Having the generic parameter in the trait means it must be named, even if all we do is add `where B: core::hash::BuildHasher`. Using an associated type with that same bound improves the ergonomics a bit.

See this example for places where we want to constrain the interpreter we get based on some of these traits in `artichoke-core`:

https://github.com/artichoke/artichoke/blob/1c1702894e6ef237b8d7644c16f6b1bc9ae5f5e7/spinoso-symbol/src/lib.rs#L234-L244

lopopolo commented on artichoke-backend/src/hash.rs:
```diff
@@ -0,0 +1,14 @@
+use crate::ffi::InterpreterExtractError;
+use crate::Artichoke;
+
+impl BuildHasher<RandomState> for Artichoke {
+    type Error = Error;
```
To match the change proposed below:

```suggestion
impl BuildHasher for Artichoke {
    type Error = Error;
    type BuildHasher = RandomState;
```

lopopolo commented on artichoke-backend/src/hash.rs:
```diff
@@ -0,0 +1,14 @@
+impl BuildHasher<RandomState> for Artichoke {
+    type Error = Error;
+
+    fn build_hasher(&mut self) -> Result<&RandomState, Self::Error> {
```
style nit, I prefer using the associated types here to mirror the trait definition:

```suggestion
    fn build_hasher(&mut self) -> Result<&Self::BuildHasher, Self::Error> {
```

lopopolo commented on artichoke-backend/src/lib.rs:
```diff
@@ -126,6 +126,7 @@ mod eval;
 pub mod exception_handler;
 pub mod extn;
 pub mod ffi;
+pub mod hash;
```
there's nothing here to export, there's only a trait implementation, so let's not make this `pub` like `globals` and `intern` below.

```suggestion
mod hash;
```

lopopolo reviewed (changes requested):
Thanks for the PR @masonforest.

It looks like I communicated poorly around the associated type parameter for the build hasher. A couple of tweaks and I think this will be good to go.

In case you're not familiar with associated types in traits, this is what they refer to:

https://doc.rust-lang.org/rust-by-example/generics/assoc_items/types.html

masonforest committed:
Apply code review feedback for `artichoke_core::hash::Hash`

Co-authored-by: Ryan Lopopolo <rjl@hyperbo.la>

masonforest committed:
Move BuildHasher to an associated type

masonforest committed:
Update documentation

masonforest commented:
Ah, I should have read more carefully. You did say to use an associated type in your feedback.

I think this should be getting close now.

lopopolo reviewed (approved):
This looks great. Thanks!

I'll merge once CI passes.

lopopolo commented:
It looks like rustfmt failed. If you run this locally it should fix everything up:

```shell
bundle install
bundle exec rake
```

Or if you don't have Ruby, it looks like all it needs is a:

```shell
cargo fmt
```

masonforest committed:
Run `cargo fmt`

masonforest committed:
Fixups for `artichoke_core::hash::Hash`

- Add #Errors section to docs.
- Group imports so `std` imports come before `crate` imports.
- Add `Hash` docs to crate README.

lopopolo committed:
Bump artichoke-core to 0.10.0

lopopolo commented:
Sorry for the back and forth here. I hope you don't mind that I've tweaked your branch to add the docs and clean things up a bit. Once the tests pass, this is good to go.

lopopolo commented:
this PR bumps `artichoke-core` to version 0.10.0.
